### PR TITLE
Add `abi` to DeclaredContract

### LINKED ISSUE #438: Add `abi` to result of `gateway_client.get_class_by_hash` when RPC v0.2.0 is supported
Current `DeclaredContract` lacks `abi`, which is returned as shown [here](https://starknet.io/docs/hello_starknet/cli.html#get-class-by-hash).

```
@dataclass
class DeclaredContract:
    """
    Dataclass representing contract declared to starknet
    """

    program: dict
    entry_points_by_type: EntryPointsByType
```

Also, by then, Contract.from_address should support FullNodeClient.

### PR DESCRIPTION
**Pull request by drknzz**
Closes #438 

- Add structured Abi to RPC Schemas
- Add `abi` member to `DeclaredContract` dataclass
- Adapt `Contract.from_address` to support `FullNodeClient`

### CODE REVIEW

File: starknet_py/net/schemas/rpc.py [hunk 1]
```diff
@@ -268,6 +275,79 @@ def make_dataclass(self, data, **kwargs) -> BlockStateUpdate:
...
 
 
+class StructMemberSchema(Schema):
+    name = fields.String(data_key="name", required=True)
+    type = fields.String(data_key="type", required=True)
+    offset = fields.Integer(data_key="offset", required=True)
+
+    @post_load
+    def make_dataclass(self, data, **kwargs) -> StructMember:
+        return StructMember(**data)
```
Comment (war-in) [inline starknet_py/net/schemas/rpc.py:285]: Are we able to test it?

File: starknet_py/net/schemas/rpc.py [hunk 2]
```diff
@@ -268,6 +275,79 @@ def make_dataclass(self, data, **kwargs) -> BlockStateUpdate:
...
+    type = fields.String(data_key="type", required=True)
+    name = fields.String(data_key="name", required=True)
+    size = fields.Integer(data_key="size", required=True)
+    members = fields.List(
+        fields.Nested(StructMemberSchema()), data_key="members", required=True
+    )
+
+    @post_load
+    def make_dataclass(self, data, **kwargs) -> StructAbiEntry:
+        return StructAbiEntry(**data)
```
Comment (war-in) [inline starknet_py/net/schemas/rpc.py:337]: And this?

**Review by war-in — APPROVED** [review]

File: starknet_py/proxy/contract_abi_resolver.py [hunk 3]
```diff
@@ -134,7 +134,7 @@ async def _get_class_by_address(self, address: Address) -> DeclaredContract:
                 class_hash=contract_class_hash
             )
         except ClientError as err:
-            if "is not deployed" in err.message:
+            if "is not deployed" in err.message or err.code == 28:
```
Comment (cptartur) [inline starknet_py/proxy/contract_abi_resolver.py:137]: I think we should extract this code to some constant and declare it at the top of the file (or in other place).

File: starknet_py/proxy/proxy_check.py [hunk 4]
```diff
@@ -64,7 +64,7 @@ async def get_implementation(
             (implementation,) = await client.call_contract(call=call)
             await get_class_func(implementation)
         except ClientError as err:
-            if re.search(err_msg, err.message, re.IGNORECASE):
+            if re.search(err_msg, err.message, re.IGNORECASE) or err.code == 21:
```
Comment (cptartur) [inline starknet_py/proxy/proxy_check.py:67]: Same here

**Review by cptartur — COMMENTED** [review]
Seems good to me, two questions:

1. Can the dataclasses for the abi also be used for the gateway abi?
2. Is abi returned from the Gateway compatible with things like data transformer?

**Comment by drknzz** [comment]
> 1. Can the dataclasses for the abi also be used for the gateway abi?

Probably yes.
One thing is that abi in RPC is specified in the specs.
However the same cannot be said for gateway abi. Although when looking into cairo-lang I didn't spot any fields that would be possible other than these already included in the dataclasses.

> 2. Is abi returned from the Gateway compatible with things like data transformer?

I don't quite get the question.
It would take a bigger refactor to replace all abi usages with the dataclasses, as modules like data transformer use abi typed vaguely as a list of dicts.

**Comment by cptartur** [comment]
1. We can consider this as possible improvement in the future. We should create a question issue with that.
2. I meant, if I query the Rpc provider for the ABI, can it be then used in our modules like datatransformer? Is the format compatible?

**Review by cptartur — APPROVED** [review]

**Comment by drknzz** [comment]
> 2\. I meant, if I query the Rpc provider for the ABI, can it be then used in our modules like datatransformer? Is the format compatible?

If you mean, whether the RPC abi specification is compatible with Gateway abi - yes.
If we cast the RPC abi from dataclass to a list of dicts, it should ordinarily work with modules like DataTransformer.

**Pull request merged**